### DISCUSSION TITLE
Capacitor / iOS / Android / NextJs - Is there a better way to deal with OAuth redirects?

### DISCUSSION THREAD

**Discussion by jpmasters** [supabase/supabase#11548] [Questions]
Hi all, I'm working on a PoC / MVP app that is built on React / NextJs and Capacitor running on iOS and Android and the app shares components with my main NextJs web app using a monorepo. 

I wanted to integrate an OAuth provider (in this case Google) and connect both the web and mobile apps to the same authentication backend using Supabase Auth (v2) and after a few days of head scratching, pouring through supabase repos, reading docs and blogs and general grumpiness (directed mostly at my own ineptitude!) I *think* I've now got it all working. There's still a bit of work I need to do for Android around signing to enable the deep linking but I think the main problems are now solved.

Integrating with the NextJs web application was very easy as the documentation covers it well. Where I had to really work on my knowledge gaps was getting it working for the mobile app. There are a few examples in blogs but it seems they either don't cover Capacitor / mobile or they don't include OAuth providers.

The way I eventually got it hooked together was to start with setting up deep links, for which there is a load of documentation out there. I'm using Universal links in iOS so redirecting to the app from my domain rather than using a custom scheme. Once I had that up and running I created a URL Listener component using URLOpenListenerEvent from `@capacitor/app` and by attaching Safari to the iOS sim I could set a breakpoint in the `useEffect` and see the redirect arriving at the app with the tokens in the query.

```typescript
import React, { useEffect } from 'react';
import { App, URLOpenListenerEvent } from '@capacitor/app';

const getParameterByName = (name: string, url?: string) => {
    if (!url) url = window?.location?.href || ''
    // eslint-disable-next-line no-useless-escape
    name = name.replace(/[\[\]]/g, '\\$&')
    const regex = new RegExp('[?&#]' + name + '(=([^&#]*)|&|#|$)'),
        results = regex.exec(url)
    if (!results) return null
    if (!results[2]) return ''
    return decodeURIComponent(results[2].replace(/\+/g, ' '))
}

const AppUrlListener: React.FC<any> = ({ onAccessTokenFound }) => {

    useEffect(() => {
        App.addListener('appUrlOpen', (event: URLOpenListenerEvent) => {
            // Example url: https://mydomain.com#access_token=...
            if (event.url.includes('#access_token=')) {

                const url = event.url

                const tokens = {
                    provider_token: getParameterByName('provider_token', url),
                    provider_refresh_token: getParameterByName('provider_refresh_token', url),
                    access_token: getParameterByName('access_token', url),
                    expires_in: getParameterByName('expires_in', url),
                    refresh_token: getParameterByName('refresh_token', url),
                    token_type: getParameterByName('token_type', url),
                }

                console.log('FOUND TOKENS', tokens)

                // callback to the app cpt
                if (onAccessTokenFound) {
                    onAccessTokenFound(tokens);
                }
            }

            // If no match, do nothing - let regular routing
            // logic take over
        });
    }, []);

    return null;
};

export default AppUrlListener;
```
I was then able to filch the getParameterByName function from gotrue-js helpers to quickly parse out the parameters and capture them into an object to pass back to the app cpt.

So far things had been pretty straightforward but the next step, I thought, should have been super simple. All I had to do was find a method in GoTrueClient that I could call to create a new session from the tokens. My app would magically sign in and all would be good however that turned out not to be the case. It looks like there was a v1 signin method that would take an access & refresh token as an input and create the new session from it but that seems to have been replace with other things in v2. I tried calling setSession and although that returns the session and all looks well, the actual sign in doesn't seem to happen and the app stubbornly refused to move past the sign in screen. 

Looking at the gotrue-js code, it looks like it'll pick up URL params in Initialize() so I tried calling that but of course, once that initializePromise is done, that's it, you can't seem to get the initialization logic to re-execute. I then tried various ways of navigating the browser to a URL with the query params intact to force the supabase client to recreate itself and initially that looked like it was going to work but alas only on javascript hot loads coming from my development machine. Once the app is picking up the code from capacitor:localhost it all stops working again. Incidentally, according to a Capacitor blog I read, Android serves the app internally using http rather than capacitor as the scheme so it might work ok for Android, I've not checked.

Finally, I got it working by using this callback function in my _app.tsx:

```typescript
const resetSupabaseClient = async (tokens: any) => {
    console.log('Need to recreate the supabase client with token', tokens)

    const res = await supabase.auth.setSession(tokens)
    console.log('SetSession called', res)

    const sess = await supabase.auth.getSession()
    console.log('getSession now returned', sess)

    const refreshRes = await supabase.auth.refreshSession(tokens)
    console.log('Called Refresh token', refreshRes)
}
```
Which is a lot nicer than reloading the whole app but still feels like a bit of a cludge due to having to call `supabase.auth.refreshSession(tokens)` to kick it all into life.

So, my questions here are:

1. Was all of this necessary or have I missed a screamingly obvious, much faster way of doing this?
2. Why is it that `setSession()` doesn't seem to do the whole job? I notice for example that it doesn't call `_notifyAllSubscribers()` which seems a bit odd to me. Is this intentional?
3. Now that it seems to work for me this may be irrelevant but should there be a way to re-run Initialize() once initializePromise has been completed? There is code in there to pick up the tokens from the URL but there doesn't seem to be a way to coax the client into doing that. Presumably this works for web apps because the authentication is done in the same browser tab so after redirecting back to the web app, a whole new client is created and the session can be created from scratch. That's not the case for mobile apps though, they're sitting there waiting to be triggered by the deep link.

Your thoughts, ideas and criticisms would be happily received!

Many thanks,
Jon

**Comment by j4w8n** [comment]
setSession should work. At least in regards to making sure the supabase client has the info needed to function as the signed-in user.

When you were only trying setSession, what behavior were you expecting to work?

**Reply by jpmasters** [reply]
Having done some digging through the code, the problem seems to be in the integration between the supabase client and auth-helpers.

I'm using the React hooks from auth-helpers/react to get hold of the session and the presence or not of that session, is used to display either the Auth UI or the app content.

The auth-helper [maintains that session in a useState hook](https://github.com/supabase/auth-helpers/blob/main/packages/react/src/components/SessionContext.tsx#L54). When the SessionContextProvider is created, it runs a useEffect to [set up an onAuthStateChange listener to monitor for changes to the session in the subabase client](https://github.com/supabase/auth-helpers/blob/main/packages/react/src/components/SessionContext.tsx#L83). When it is notified of that change, [it calls its own setSession mutator](https://github.com/supabase/auth-helpers/blob/main/packages/react/src/components/SessionContext.tsx#L88) which will cause React to re-render the relevant components.

However, in [gotrue-js in the setSession() function](https://github.com/supabase/gotrue-js/blob/master/src/GoTrueClient.ts#L640) you can see that _notifyAllSubscribers() isn't called. Therefore SessionContextProvider never gets to hear about it, doesn't call setSession and react doesn't re-render anything.

If you look at gotrue-js refreshSession() method by contrast, it calls the private function _callRefreshToken() which [does call _notifyAllSubscribers](https://github.com/supabase/gotrue-js/blob/master/src/GoTrueClient.ts#L1018) and I think that's probably why including the call to refreshSession() makes it work.

**Reply by j4w8n** [reply]
Ah, roger that. Nice work! Yeah refreshSession fires off the TOKEN_REFRESHED event.

I'm seeing more and more need for setSession to fire the SIGNED_IN event. I'll create a PR with some justification and links. Then we'll see what the team thinks.

**Reply by jpmasters** [reply]
No worries, happy to help. I did see another discussion [here](https://github.com/supabase/supabase/discussions/10833) where the poster took a different approach and called the notify function himself but it seemed a bit risky to me to be calling a private function. Too easy for a refactoring session to change its scope.

**Reply by j4w8n** [reply]
The PR - https://github.com/supabase/gotrue-js/pull/581
